platform/android/src/jni.cpp

```diff
@@ -1620,6 +1753,363 @@ void JNICALL nativeRemoveCustomLayer(JNIEnv *env, jobject obj, jlong nativeMapVi
+        if (renderDetach) {
+            mbgl::Log::Debug(mbgl::Event::JNI, "Attached.");
+        }
+
+        if (error) {
+            std::string message = mbgl::util::toString(error);
+            env2->CallVoidMethod(listCallback, listOnErrorMethodId, std_string_to_jstring(env2, message));
+        } else if (regions) {
+            // Build the OfflineManager object
+            jobject jmanager = env2->NewObject(offlineManagerClass, offlineManagerClassConstructorId);
+            env2->SetLongField(jmanager, offlineManagerClassPtrId, defaultFileSourcePtr);
+
+            // Build jobjectArray
+            jsize index = 0;
+            jobjectArray jregions = env2->NewObjectArray(regions->size(), offlineRegionClass, NULL);
+            for (std::vector<mbgl::OfflineRegion>::iterator region = regions->begin(); region != regions->end(); ++region) {
+                // Build the Region object
+                jobject jregion = env2->NewObject(offlineRegionClass, offlineRegionConstructorId);
+                env2->SetObjectField(jregion, offlineRegionOfflineManagerId, jmanager);
+                env2->SetLongField(jregion, offlineRegionIdId, region->getID());
+                env2->SetObjectField(jregion, offlineRegionMetadataId, metadata_from_native(env2, region->getMetadata()));
+
+                // Moves the region on the stack into a heap-allocated one
+                env2->SetLongField(jregion, offlineRegionPtrId,
+                    reinterpret_cast<jlong>(new mbgl::OfflineRegion(std::move(*region))));
+
+                env2->SetObjectArrayElement(jregions, index, jregion);
+                index++;
+            }
+
+            // Trigger callback
+            env2->CallVoidMethod(listCallback, listOnListMethodId, jregions);
+        }
+
+        // Detach when we're done
+        detach_jni_thread(theJVM, &env2, renderDetach);
+    });
+}
+
+void JNICALL createOfflineRegion(JNIEnv *env, jobject obj, jlong defaultFileSourcePtr, jobject definition_, jobject metadata_, jobject createCallback) {
+    mbgl::Log::Debug(mbgl::Event::JNI, "createOfflineRegion");
+
+    // Checks
+    assert(defaultFileSourcePtr != 0);
+    if (createCallback == nullptr) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Callback has to be set.");
+        return;
+    }
+
+    // Definition fields
+    jstring jStyleURL = reinterpret_cast<jstring>(env->GetObjectField(definition_, offlineRegionDefinitionStyleURLId));
+    std::string styleURL = std_string_from_jstring(env, jStyleURL);
+    jobject jBounds = env->GetObjectField(definition_, offlineRegionDefinitionBoundsId);
+    jdouble jMinZoom = env->GetDoubleField(definition_, offlineRegionDefinitionMinZoomId);
+    jdouble jMaxZoom = env->GetDoubleField(definition_, offlineRegionDefinitionMaxZoomId);
+    jfloat jPixelRatio = env->GetFloatField(definition_, offlineRegionDefinitionPixelRatioId);
+
+    // Bounds fields
+    jdouble mLatNorth = env->GetDoubleField(jBounds, latLngBoundsLatNorthId);
```

Comment:
Method extracted in https://github.com/mapbox/mapbox-gl-native/commit/0cc6a9b68668a9d10d645fc389233435b074a382.